feat(isTaxID): Accept ISO 3166-1 alpha-2 country codes

bartmalanczuk opened the pull request:
Changes accepted locale format for `isTaxID` validator

Resolves #1707 

## Checklist

- [x] PR contains only changes related; no stray files, etc.
- [x] README updated (where applicable)
- [x] Tests written (where applicable)

bartmalanczuk committed:
Accept ISO 3166-1 alpha-2 country codes in isTaxID

bartmalanczuk committed:
Order locales alphabetically

tux-tn reviewed (commented):
Hello @bartmalanczuk thank you for your PR.
This is a breaking change to `isTaxID`. We should probably find a way to handle correctly the transition without impacting the current behaviour.

fedeci commented:
Aliases everywhere🙃

bartmalanczuk commented:
@tux-tn understood- I can provide a fix supporting both variants

bartmalanczuk committed:
Support previous locale format for backward compatibility

tux-tn reviewed (commented):
Honestly i didn't see it that way, was thinking for an easier and less naive appraoch:
```javascript
const legacyIsoLocales = new Set([ 'bg-BG', 'cs-CZ', 'de-AT', 'de-DE', 'dk-DK', 'el-CY', 'el-GR', 'en-GB', 'en-IE', 'en-US', 'es-ES', 'et-EE', 'fi-FI', 'fr-BE', 'fr-FR', 'fr-LU', 'hr-HR', 'hu-HU', 'it-IT', 'lb-LU', 'lt-LT', 'lv-LV', 'mt-MT', 'nl-BE', 'nl-NL', 'pl-PL', 'pt-BR', 'pt-PT', 'ro-RO', 'sk-SK', 'sl-SI', 'sv-SE']);

export default function isTaxID(str, locale = 'US') {
  assertString(str);
  // Copy TIN to avoid replacement if sanitized
  let strcopy = str.slice(0);

  if (legacyIsoLocales.has(locale)) {
    locale = locale.substring(3); // strip ISO 639-1 language code if a legacy locale is provided
  }

  if (locale in taxIdFormat) {
    if (locale in sanitizeRegexes) {
      strcopy = strcopy.replace(sanitizeRegexes[locale], '');
    }
    if (!taxIdFormat[locale].test(strcopy)) {
      return false;
    }

    if (locale in taxIdCheck) {
      return taxIdCheck[locale](strcopy);
    }
    // Fallthrough; not all locales have algorithmic checks
    return true;
  }
  throw new Error(`Invalid locale '${locale}'`);
}
```
cc @fedeci what's your opinion about this? should we duplicate all elements or just remove language code from locale before testing?

bartmalanczuk committed:
Refactor legacy aliases

bartmalanczuk committed:
Merge remote-tracking branch 'origin/master' into accept-iso-3166-1-alpha-2-country-codes-in-istaxid

tux-tn reviewed (changes requested):
Thank you for making the changes @bartmalanczuk ! I added some comments (I promise it will be the last ones 😅  ) 
Can you please address them?

bartmalanczuk committed:
Revert reordering aliases

bartmalanczuk committed:
Test legacy aliases

bartmalanczuk committed:
Test legacy aliases

bartmalanczuk commented:
@tux-tn I reverted reordering (but I kept renaming of the functions) and simplified the tests as requested. Can you only advise me what to do with the Readme/ documentation because I'm not sure if we are on the same page 😅

clytras commented:
Thanks for the awesome work @bartmalanczuk and @tux-tn.
Please merge!

tux-tn commented:
cc @profnandaa any feedback on this?
